### PR TITLE
Implementing SSRS – Substrate Simple Remote Signer protocol

### PR DESCRIPTION
**Pull request by gnunicorn**
Implementation of the jsonrpc remote keystore over the http/ws protocol as the substrate keystore, including a separate example server implementation. 

To Do:
- [x] docs
- [x] Metadata, License
- [x] hide behind non-default feature-flag
- [x] implement remaining (currently unused) functions on server

_Note_: test had ecdsa support, but was removed because of https://github.com/paritytech/substrate/issues/7490

### CODE REVIEW

**Comment by NikVolf** [comment]
> over the http/ws

Shouldn't it be https/wss? How do you secure the transport anyway?

**Comment by gnunicorn** [comment]
> Shouldn't it be https/wss? How do you secure the transport anyway?

This is using regular [jsonrpc_client_transports](https://docs.rs/jsonrpc-client-transports/15.1.0/jsonrpc_client_transports/transports/index.html), which only support `http` and `ws` afaics.

**Comment by NikVolf** [comment]
> > Shouldn't it be https/wss? How do you secure the transport anyway?
> 
> This is using regular [jsonrpc_client_transports](https://docs.rs/jsonrpc-client-transports/15.1.0/jsonrpc_client_transports/transports/index.html), which only support `http` and `ws` afaics.

it is not a secure transport, especially for this task

**Comment by gnunicorn** [comment]
> it is not a secure transport, especially for this task

That is correct, and no one said it was. The expected environment for this to this point was to have a secured network, like run in a VPN and firewalled system and/or expect the connection to be tunneled through external means.

I see though that the regular user might not realize they have to do that, so I could see that we hide that feature behind a build flag you'd have to explicitly activate until we have a secure-default-way of doing it instead. WDYT?

**Comment by NikVolf** [comment]
> > it is not a secure transport, especially for this task
> 
> That is correct, and no one said it was. The expected environment for this until now, was to have a secured network, like run in a VPN and firewalled sub-network and/or expect the connection to be tunneled through external means (like SSH).
> 
> I see though that the regular user might not realize they have to do that, so I could see that we might want hide that this behind a feature flag you'd have to explicitly activate until we have a secure-default-way of doing it instead. WDYT?

Signer will be custom software implemented anyway. So I don't think there is much sense in using non-secure transport at all, when there are secure alternatives (for example, [noise](https://github.com/mcginty/snow)). If it is properly used, there will be no need in tunnelling and stuff. Noise has implementation in many languages if that is what required - even in pure `C`.

JSON-RPC overhead need is also not clear to me, but I am not sure if this is critical. How much signing happens on validator node per second/minute?

**Comment by gnunicorn** [comment]
@NikVolf the goal was to get something out of the door quickly with the least developer overhead from our side to allow others to start experimenting with the workflow and use case – JSONRPC was all there and quick to get up and running, that's the only reason it was used. The way it stands, there are known security problems (like, the server still needs to have some knowledge of chain-info otherwise they can't securely sign VRFs, but that also means they have _some_ connection to a chain or chain state, which requires internet to stay up to date) and performance issues (https and JSON being two things that could be replaced with more efficient things instead, but also that signing is mostly sync for in substrate still and the example and known others approaches also follow linear service patterns – like TEE can only do one signing at a time).

**Comment by burdges** [comment]
I've a stupid question:  I'd think half the tools for doing JSON RPC should do TLS, no?

**Comment by kirushik** [comment]
After conversation with @gnunicorn I've removed the `needsaudit` flag; our current SSRS implementation is for experimentation only, hidden behind a feature flag and should under no circumstances be used in anything production-related.

@gnunicorn is extending the readme to explicitly state the above.

(This shouldn't stop us from raising and documenting security-related issues with it, though — it's just by far not in shape to sustain any _professional attention_ yet, and probably never will in its current form.)

**Comment by bkchr** [comment]
> After conversation with @gnunicorn I've removed the `needsaudit` flag; our current SSRS implementation is for experimentation only, hidden behind a feature flag and should under no circumstances be used in anything production-related.
> 
> @gnunicorn is extending the readme to explicitly state the above.
> 
> (This shouldn't stop us from raising and documenting security-related issues with it, though — it's just by far not in shape to sustain any _professional attention_ yet, and probably never will in its current form.)

Maybe we should than put this into its own repo? 

And what is the path we continue here? I thought this issue is soo pressing that we can deprecate sentries?

**Comment by kirushik** [comment]
> And what is the path we continue here?

I must admit, I'm a bit confused here too.
I'm talking to @gnunicorn in private to better understand what should and what shouldn't go into our codebase/pass the review here.

**Comment by gnunicorn** [comment]
Convertet this back to draft to more clearly signal:
- we don't the intend to merge or ship this as-is
- this is an example implementation showing how in the new API other remote-protocols can easily be integrated
- this is a highly insecure-by-default-example and before any of it could be used, must be given appropriate documentation

Unfortunately, no other, more secure protocol has been proposed until now, which leaves us with a couple of path to go ahead, but not decided on any yet:
- leave this as is and wait for other protocols to come around
- replace the given transport with an at least TLS-wrapped one and probably some further hardening, like certificate pinning or off-protocol key-exchange-requirement – none of which is currently supported by the libraries used
- provide a generic protocol-plugging-mechanism and move the specific implementation into a example/experimental-named repository to make clear this is more about showing the API then the specific protocol

**Comment by burdges** [comment]
I'd assumed remote signers were vaguely HSM-like with special hardware connections, like a second ethernet card or USB connection.  Yes, I think an internet remote signer wants a secure handshake like TLS but with fixed transport keys and IP addresses for both sides, so no CAs or DNS.

Interestingly, if you look at Noise IK and KK then you'll observe the difference is sending the initiators long term secret inside the first flow, encrypted only by an ephemeral key exchange. 
```
IK:
  <- s
  ...
  -> e, es, s, ss
  <- e, ee, se,

KK:
  -> s
  <- s
  ...
  -> e, es, ss
  <- e, ee, se,
```
As the initiator knows the respondents long-term secret, one could've a hybrid handshake between IK and KK in which foreign initiators reveal their identity, which permits whitelisting on the first flow, but special distinguished initiators hide their identity against an adversary who compromises the respondents' IP address but not the respondent machine.  In other words, your remote signer can open connections that look exactly like normal node connections, but avoids immediately exposing your remote singer's IP address against BGP attacks, etc.  This is probably useless since a signer's messages look different anyways, but technically the traffic observing attacker looks different form the BGP attacker.

**Comment by redzsina** [comment]
For now, we consider our review done for this PR - please refer to issue 21 in the internal tracker.

File: client/service/src/builder.rs [hunk 1]
```diff
@@ -205,12 +208,27 @@ pub type TLightClientWithBackend<TBl, TRtApi, TExecDisp, TBackend> = Client<
 	TRtApi,
 >;
 
-enum KeystoreContainerInner {
-	Local(Arc<LocalKeystore>)
+trait AsCryptoRef {
```
Comment (rakanalh) [inline client/service/src/builder.rs:211]: Nitpick: can this be renamed to `AsCryptoStoreRef`?

File: client/service/src/builder.rs [hunk 2]
```diff
@@ -205,12 +208,27 @@ pub type TLightClientWithBackend<TBl, TRtApi, TExecDisp, TBackend> = Client<
...
+    fn keystore_ref(&self) -> Arc<dyn CryptoStore>;
+    fn sync_keystore_ref(&self) -> Arc<dyn SyncCryptoStore>;
+}
+
+struct DoubleWrap<T>(Arc<T>);
+
+impl<T> AsCryptoRef for DoubleWrap<T> where T: CryptoStore + SyncCryptoStore + 'static {
```
Comment (rakanalh) [inline client/service/src/builder.rs:218]: Question: What happens if we get rid of `DoubleWrap` and implement `AsCryptoRef` on `Arc<T>` directly? the below methods could return `self.clone()` in that case and the code would be a bit simpler.

I have modified this snippet you showed me earlier to do this:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5d6a9deb6c38c00cab9d0198b95d690c

File: client/service/src/builder.rs [hunk 3]
```diff
@@ -205,12 +208,27 @@ pub type TLightClientWithBackend<TBl, TRtApi, TExecDisp, TBackend> = Client<
...
+    }
+    fn sync_keystore_ref(&self) -> Arc<dyn SyncCryptoStore> {
+        self.0.clone()
+    }
 }
 
 /// Construct and hold different layers of Keystore wrappers
-pub struct KeystoreContainer(KeystoreContainerInner);
+pub struct KeystoreContainer {
+	remotes: Vec<Box<dyn AsCryptoRef>>,
```
Comment (rakanalh) [inline client/service/src/builder.rs:229]: If we plan to support multiple remotes, it would be nice to document the use case.

**Comment by gnunicorn** [comment]
superseeded by #7628 .

**Pull request closed without merging**